### PR TITLE
[SLM] Android refactor for SLM

### PR DESCRIPTION
**Pull request by Kartik14**
The recent SLM changes broke the compatibility with the Android app code. This PR address this by:

- ~~Add the `model_lib` and `local_id` fields to the `mlc-chat-config.json`, which are used by the android code for loading model~~
- ~~change the naming logic for `system-lib-prefix` for ios/android libs during compilation~~
- Add `model_lib` and `model_lib_path` fields to `app-config.json` to remove dependency on `mlc-chat-config.json`
- Remove `add model by url` from android app since it causes confusion
- Update the android documentation

~~This PR will require the configs on [huggingface mlc-ai models](https://huggingface.co/mlc-ai?message=Request%20sent) to be updated to include the new fields otherwise the app will crash when trying to read the `model_lib` and `local_id` fields  from current config files.~~

cc @davidpissarra @tqchen

### CODE REVIEW

**Comment by CharlieFRuan** [comment]
Thanks for pushing on Android! IIUC, one of the goals for SLM is to remove the concept of `model_lib` and `local_id` to make things more direct.

I am not too familiar with Android's workflow, but is it possible to manually enter `model_lib` field in [app-config.json](https://github.com/mlc-ai/mlc-llm/blob/main/android/library/src/main/assets/app-config.json) so that we do not need `local_id` and `model_lib` in `mlc-chat-config.json`? This change was needed for web-llm in [this PR](https://github.com/mlc-ai/web-llm/pull/240) as well for your reference.

**Comment by CharlieFRuan** [comment]
Would be great if you could populate the Android part of Prebuilt Model page here as well https://github.com/mlc-ai/mlc-llm/blob/main/docs/prebuilt_models.rst?plain=1#L213-L244. It is currently commented out as I was not sure if it still up to date with SLM. Some quick directions on how to run a prebuilt model on Android would be enough, then we can let users read more in the actual ``android.rst``. Thanks!

**Comment by Kartik14** [comment]
> Thanks for pushing on Android! IIUC, one of the goals for SLM is to remove the concept of `model_lib` and `local_id` to make things more direct.
> 
> I am not too familiar with Android's workflow, but is it possible to manually enter `model_lib` field in [app-config.json](https://github.com/mlc-ai/mlc-llm/blob/main/android/library/src/main/assets/app-config.json) so that we do not need `local_id` and `model_lib` in `mlc-chat-config.json`? This change was needed for web-llm in [this PR](https://github.com/mlc-ai/web-llm/pull/240) as well for your reference.

**Comment by Kartik14** [comment]
> Thanks for pushing on Android! IIUC, one of the goals for SLM is to remove the concept of `model_lib` and `local_id` to make things more direct.
> 
> I am not too familiar with Android's workflow, but is it possible to manually enter `model_lib` field in [app-config.json](https://github.com/mlc-ai/mlc-llm/blob/main/android/library/src/main/assets/app-config.json) so that we do not need `local_id` and `model_lib` in `mlc-chat-config.json`? This change was needed for web-llm in [this PR](https://github.com/mlc-ai/web-llm/pull/240) as well for your reference.

Thanks, I will take a look at this

**Comment by Kartik14** [comment]
> Would be great if you could populate the Android part of Prebuilt Model page here as well https://github.com/mlc-ai/mlc-llm/blob/main/docs/prebuilt_models.rst?plain=1#L213-L244. It is currently commented out as I was not sure if it still up to date with SLM. Some quick directions on how to run a prebuilt model on Android would be enough, then we can let users read more in the actual `android.rst`. Thanks!

Yeah, sounds good. I will add the instructions

**Comment by Kartik14** [comment]
Updated the code for the new SLM changes @CharlieFRuan

File: docs/prebuilt_models.rst [hunk 1]
```diff
@@ -210,40 +210,40 @@ See the full list of all precompiled model libraries `here <https://github.com/m
...
+    * - Model code
+      - Model Series
+      - Quantization Mode
+      - Hugging Face repo
+    * - `Llama-2-7b-q4f16_1`
+      - `Llama <https://ai.meta.com/llama/>`__
+      - * Weight storage data type: int4
+        * Running data type: float16
+        * Symmetric quantization
+      - `link <https://huggingface.co/mlc-ai/mlc-chat-Llama-2-7b-chat-hf-q4f16_1>`__
```
Comment (CharlieFRuan) [inline docs/prebuilt_models.rst:237]: This should probably be `https://huggingface.co/mlc-ai/Llama-2-7b-chat-hf-q4f16_1-MLC` instead; and `https://huggingface.co/mlc-ai/RedPajama-INCITE-Chat-3B-v1-q4f16_1-MLC` for RedPajama

File: docs/deploy/android.rst [hunk 2]
```diff
@@ -85,43 +85,45 @@ To deploy models on Android with reasonable performance, one has to cross-compil
...
-          --model ./dist/models/$MODEL_NAME \
-          --quantization $QUANTIZATION
+  # convert weights
+  mlc_chat convert_weight ./dist/models/$MODEL_NAME/ --quantization $QUANTIZATION -o dist/$MODEL_NAME-$QUANTIZATION-MLC/
+  # create mlc-chat-config.json
+  mlc_chat gen_config ./dist/models/$MODEL_NAME/ --quantization $QUANTIZATION \
+    --conv-template llama-2 --context-window-size 768 -o dist/${MODEL_NAME}-${QUANTIZATION}-MLC/
+  # 2. compile: compile model library with specification in mlc-chat-config.json
+  mlc_chat compile ./dist/${MODEL_NAME}-${QUANTIZATION}-MLC/mlc-chat-config.json \
+      --device android -o ./dist/${MODEL_NAME}-${QUANTIZATION}-MLC/${MODEL_NAME}-${QUANTIZATION}-MLC-android.tar
```
Comment (CharlieFRuan) [inline docs/deploy/android.rst:98]: Let's perhaps remove `-MLC` in model libs (keep in weights) to make it consistent across platforms (lines 105 and 126 as well); makes the names shorter too

**Review by CharlieFRuan — COMMENTED** [review]
Thank you so much for the changes! Some small nit about namings; otherwise LGTM.

**Comment by CharlieFRuan** [comment]
Would be great if we could upload Android's prebuilt model library too to finish the last mile: https://github.com/mlc-ai/binary-mlc-llm-libs/tree/main/Llama-2-7b-chat-hf (and redpajama).

We could add a new column in:
https://github.com/mlc-ai/mlc-llm/blob/7684c07226832384a89990aee55301ef92a5da31/docs/prebuilt_models.rst?plain=1#L324

**Comment by Kartik14** [comment]
> Would be great if we could upload Android's prebuilt model library too to finish the last mile: https://github.com/mlc-ai/binary-mlc-llm-libs/tree/main/Llama-2-7b-chat-hf (and redpajama). Thank you!!
> 
> We could add a new column in:
> 
> https://github.com/mlc-ai/mlc-llm/blob/7684c07226832384a89990aee55301ef92a5da31/docs/prebuilt_models.rst?plain=1#L324

I have made a PR adding these models, once it is merged, I will create a new column for android models

**Comment by CharlieFRuan** [comment]
Merged the libraries in!

**Comment by CharlieFRuan** [comment]
@Kartik14 Shall we merge this in?

**Pull request merged**